doc/man/ejsc.1
.TH EC "1" "March 2009" "ejsc" "User Commands"
.SH NAME
ejsc \- Ejscript compiler.
.SH SYNOPSIS
.B ejsc
[\fI--bind\fR]
[\fI--debug\fR]
[\fI--doc\fR]
[\fI--lang [ecma|plus|fixed]\fR]
[\fI--empty\fR]
[\fI--noout\fR]
[\fI--optimize level\fR]
[\fI--out filename\fR]
[\fI--parse\fR]
[\fI--searchPath ejsPath\fR]
[\fI--standard\fR]
[\fI--strict\fR]
[\fI--use 'module, ...'\fR]
[\fI--version\fR]
[\fI--warn level\fR]
files...
.SH DESCRIPTION
The \fBejsc\fR command compiles Ejscript programs to produce Ejscript module files containing byte code.
The module files may then be run via the \fBejsvm\fR virtual machine command or via the \fBejs\fR shell command.
.PP
The \fBejsc\fR command is both a compiler and link editor. Inputs can be either Ejscript source files or
Ejscript modules that have come from previous invocations of the \fBejsc\fR command. Outputs will be one or more Ejscript 
module files that contain declarations and byte code. 
.PP
An output module file will be created for each Ejscript module directive
encountered during compilation. A module file will also be created for any global variables or functions 
declared outside a module directive. These global declarations will go into the \fBdefault\fR module that is 
specially reserved for global declarations.  Each module file will be named according to the module directive name, 
but with a \fB.mod\fR extension appended. 
.SH LINKING
If \fBejsc\fR is invoked with the --out switch, all input scripts, modules and any other dependent modules are merged together 
into a single output module. The modules retain their logical naming, but are emitted in a single output module file. When 
that module file is loaded, all the contained modules will be available to the program. This creates not only a convenient 
way to package an entire application as a single file, it also permits many optimizations by merging the entire application 
and its dependent modules into a single module.
.SH BINDING
The \fBejsc\fR compiler will attempt to early-bind all possible variable and function references. Binding means resolving 
references to the underlying storage for properties and functions. Doing this at compile time usually results in much
faster execution at run-time.  
.PP
When using the --out switch, the compiler can early-bind all global variables, functions and 
type references, resulting in a much smaller and faster application. However, you must not subsequently load other 
modules that also have global declarations. Otherwise the Ejscript loader will throw an exception. Consequently, 
the --out switch must only be used to create a complete application including all the application's required modules.
.SH COMPLIANCE
Ejscript is fully compliant with the JavaScript (ECMA-262 3.X) standards, but it also offers a set of 
enhancements and fixes designed to improve the language and correct some longstanding JavaScript design issues. 
Use of these enhancements and fixes is optional and controlled via opt-in configuration switches and script pragmas.
.PP
The \fBejsc\fR compiler supports three language compliance modes: 
.IP "" 4
* \fBecma\fR  for strict ECMAScript compliance.
.IP "" 4
* \fBplus\fR  for close compliance plus compatible Ejscript enhancements.
.IP "" 4
* \fBfixed\fR  for close compliance, plus compatible Ejscript enhancements and breaking fixes.
.PP
When Ejscript is built, the default compliance mode is defined. By default, this is \fBplus\fR mode.
The \fBfixed\fR mode corrects several issues with Javascript that remain in the language due to browser compatibility 
requirements. Ejscript, by targeting non-browser envirnonments, can rectify these issues without impact to 
legacy applications. These changes are:
.IP "" 4
* Assignments to non-existent properties inside functions, will create local variables rather than global variables.
.IP "" 4
* Multiple declarations of the same variable in the same scope are not permitted.
.IP "" 4
* The == and != operators will perform like their more rigorous conterparts === and !===.
.SH OPTIONS
.TP
\fB\--bind\fR
Bind global variable declarations into slot references. This results in faster more compact code. However, only
one module can have its global variables bound.
.TP
\fB\--debug\fR
Generate symbolic debug instructions. This permits symbolic debugging of Ejscript programs and enables exception
stack backtraces to include line number information.
.TP
\fB\--doc\fR
Include documentation strings from input scripts in the output modules. The \fBejsmod\fR command can then generate
HTML documentation using these doc strings. The format of the doc strings resembles that of Javadoc.
.TP
\fB\--empty\fR
Start with an empty interpreter without the core language types such as Object, Array and Number. This option is 
used to build the foundation ejs.mod module which contains the core system types.
.TP
\fB\--lang [ecma|plus|fixed]\fR
Set the language compliance mode. Use \fBecma\fR for strict ECMAScript Edition 3 compliance. Use \fBplus\fR for 
close ECMAScript compliance plus Ejscript enhancements. Select \fBfixed\fR mode for ECMAScript features plus
enhancements and some compatibility breaking fixes.
.TP
\fB\--merge\fR
Merge all input files and modules together into a single output module. This option acts like a link-editor combining
all inputs together. Useful if you want to distribute your application as a single module file.
.TP
\fB\--optimize level\fR
Set the code optimization level. Level values must be between 0 (least) and 9 (most). Default is 9.
.TP
\fB\--parse\fR
Just parse the source scripts. Don't verify, execute or generate output. Useful to check the script syntax only.
.TP
\fB\--searchPath ejsPath\fR
Override the module search path. The module search path is a set of directories that the \fBejsc\fR command will use
when locating and loading Ejscript modules. Given a module named "a.b.c" in a script, \fBejsc\fR will use the following 
search strategy to locate the module:
.IP
1. Search for a module file named "a.b.c.mod"
.IP
2. Search for a module file named "a/b/c.mod"
.IP
3. Search for a module file named "a.b.c.mod" in EJSPATH
.IP
4. Search for a module file named c.mod in EJSPATH
.IP
The search path is initially specified by the environment variable EJSPATH and may be overridden 
by using the --searchPath ejsPath switch. EJSPATH and the ejsPath command line value are similar to 
the system PATH formats. On windows, path segments 
are separated by ";" and on Linux, Unix, FreeBSD and MAC, the path segments are separated by ":" delimiters.
.TP
\fB\--standard\fR
Run scripts in standard mode. Ejscript supports two parsing modes: strict and standard. Standard mode does not require 
variables be declared and typed before use.
.TP
\fB\--strict\fR
Run scripts in standard mode. Ejscript supports two parsing modes: strict and standard. Strict mode requires that 
all variables be declared and typed.
.TP
\fB\--use 'module, ...'\fR
List of modules to preload before compiling input files.
.TP
\fB\--version\fR
Print the \fBejsc\fR command version and exit.
.TP
\fB\--warn level \fR
Set the compiler warning verbosity level. Level values must be between 0 (least verbose) and 9 (most). Default is 0.
.SH BACKGROUND
Ejscript is an enhanced implementation of the JavaScript language for use in embedded applications such as web servers, 
embedded and mobile devices. It is especially suited for Server-Side JavaScript web applications.
.PP
Ejscript is a dynamic, interpreted, object-oriented scripting language that supports classes, objects, 
exceptions, statements, expressions and a powerful suite of data types.
.SH "REPORTING BUGS"
Report bugs to dev@embedthis.com.
.SH COPYRIGHT
Copyright \(co 2004-2009 Embedthis Software.  Ejscript is a trademark of Embedthis Software.
.br
.SH "SEE ALSO"
ejsgen, ejs, ejsmod, ejsvm


src/cmd/ejsc.c
/**
 *  ejsc.c - Ejscript Compiler main program
 *
 *  This compiler will compile the files given on the command line.
 *
 *  Copyright (c) All Rights Reserved. See details at the end of the file.
 */

/********************************** Includes **********************************/

#include    "ejs.h"
#include    "ecCompiler.h"

/********************************** Forwards **********************************/

static int preloadModules(EcCompiler *cp, MprList *modules);

/************************************ Code ************************************/

int main(int argc, char **argv)
{
    Mpr             *mpr;
    Ejs             *ejs;
    EcCompiler      *cp;
    EjsService      *vmService;
    MprList         *useModules;
    char            *argp, *searchPath, *outputFile, *certFile, *name, *tok, *modules, *spec, *order;
    int             nextArg, err, ejsFlags, ecFlags, bind, debug, doc, empty, merge, nobind, endian;
    int             warnLevel, noout, parseOnly, tabWidth, optimizeLevel, compilerMode, strip, lang;

    /*
     *  Create the Embedthis Portable Runtime (MPR) and setup a memory failure handler
     */
    mpr = mprCreate(argc, argv, ejsMemoryFailure);
    mprSetAppName(mpr, argv[0], 0, 0);

    if (mprStart(mpr, 0) < 0) {
        mprError(mpr, "Can't start mpr services");
        return EJS_ERR;
    }

    err = 0;
    searchPath = 0;
    ejsFlags = EJS_FLAG_COMPILER | EJS_FLAG_NO_EXE;
    compilerMode = PRAGMA_MODE_STANDARD;
    certFile = 0;
    ecFlags = 0;

    bind = 0;
    debug = 0;
    doc = 0;
    empty = 0;
    merge = 0;
    nobind = 0;
    noout = 0;
    parseOnly = 0;
    tabWidth = 4;
    warnLevel = 1;
    outputFile = 0;
    optimizeLevel = 9;
    lang = BLD_FEATURE_EJS_LANG;

    endian = mprGetEndian(mpr);
    useModules = mprCreateList(mpr);

    for (nextArg = 1; nextArg < argc; nextArg++) {
        argp = argv[nextArg];
        if (*argp != '-') {
            break;
        }
        if (strcmp(argp, "--bind") == 0) {
            bind = 1;

        } else if (strcmp(argp, "--cross") == 0) {
            endian = BLD_HOST_ENDIAN;

        } else if (strcmp(argp, "--debug") == 0) {
            debug = 1;

        } else if (strcmp(argp, "--doc") == 0) {
            doc = 1;

        } else if (strcmp(argp, "--lang") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                spec = argv[++nextArg];
                if (mprStrcmpAnyCase(spec, "ecma") == 0) {
                    lang = EJS_SPEC_ECMA;
                } else if (mprStrcmpAnyCase(spec, "plus") == 0) {
                    lang = EJS_SPEC_PLUS;
                } else if (mprStrcmpAnyCase(spec, "fixed") == 0) {
                    lang = EJS_SPEC_FIXED;
                }
            }

        } else if (strcmp(argp, "--empty") == 0) {
            empty = 1;

        } else if (strcmp(argp, "--endian") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                order = argv[++nextArg];
                if (mprStrcmpAnyCase(order, "big") == 0) {
                    endian = MPR_BIG_ENDIAN;
                } else if (mprStrcmpAnyCase(order, "little") == 0) {
                    endian = MPR_LITTLE_ENDIAN;
                } else {
                    err++;
                }
            }

        } else if (strcmp(argp, "--log") == 0) {
            /*
             *  Undocumented logging switch
             */
            if (nextArg >= argc) {
                err++;
            } else {
                ejsStartLogging(mpr, argv[++nextArg]);
            }

        } else if (strcmp(argp, "--merge") == 0) {
            merge = 1;

        } else if (strcmp(argp, "--nobind") == 0) {
            /*
             *  This is a hidden switch just for the compiler developers
             */
            nobind = 1;

        } else if (strcmp(argp, "--noout") == 0) {
            noout = 1;

        } else if (strcmp(argp, "--standard") == 0) {
            compilerMode = PRAGMA_MODE_STANDARD;

        } else if (strcmp(argp, "--strict") == 0) {
            compilerMode = PRAGMA_MODE_STRICT;

        } else if (strcmp(argp, "--optimize") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                optimizeLevel = atoi(argv[++nextArg]);
            }

        } else if (strcmp(argp, "--out") == 0) {
            /*
             *  Create a single output module file containing all modules
             */
            if (nextArg >= argc) {
                err++;
            } else {
                outputFile = argv[++nextArg];
            }

        } else if (strcmp(argp, "--parse") == 0) {
            parseOnly = 1;

        } else if (strcmp(argp, "--searchpath") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                searchPath = argv[++nextArg];
            }

        } else if (strcmp(argp, "--sign") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                certFile = argv[++nextArg];
            }

        } else if (strcmp(argp, "--strip") == 0) {
            strip = 1;

        } else if (strcmp(argp, "--tabWidth") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                tabWidth = atoi(argv[++nextArg]);
            }

        } else if (strcmp(argp, "--use") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                modules = mprStrdup(mpr, argv[++nextArg]);
                name = mprStrTok(modules, " \t", &tok);
                while (name != NULL) {
                    mprAddItem(useModules, name);
                    name = mprStrTok(NULL, " \t", &tok);
                }
            }

        } else if (strcmp(argp, "--version") == 0 || strcmp(argp, "-V") == 0) {
            mprErrorPrintf(mpr, "%s %s\n"
                "Copyright (C) Embedthis Software 2003-2009\n"
                "Copyright (C) Michael O'Brien 2003-2009\n",
               BLD_NAME, BLD_VERSION);
            exit(0);

        } else if (strcmp(argp, "--warn") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                warnLevel = atoi(argv[++nextArg]);
            }

#if BLD_FEATURE_EJS_WEB
        } else if (strcmp(argp, "--web") == 0) {
#if BLD_FEATURE_EJS_DB
            mprAddItem(useModules, "ejs.db");
#endif
            mprAddItem(useModules, "ejs.web");
#endif

        } else {
            err++;
            break;
        }
    }

    if (noout || merge) {
        bind = 1;
    }
    if (nobind) {
        bind = 0;
    }

    if (outputFile && noout) {
        mprErrorPrintf(mpr, "Can't use --out and --noout\n");
        err++;
    }

    if (argc == nextArg) {
        err++;
    }

    if (err) {
        /*
         *  Usage Examples:
         *      ejsc Person.es User.es Customer.es
         *      ejsc --out group.mod Person.es User.es Customer.es
         *      ejsc --out group.mod Person.es User.es Customer.es
         */
        mprErrorPrintf(mpr,
            "Usage: %s [options] files...\n"
            "  Ejscript compiler options:\n"
            "  --bind               # Bind global properties to slots. Requires --out.\n"
            "  --cross              # Cross compilation. Will use host system byte ordering.\n"
            "  --debug              # Include symbolic debugging information in output\n"
            "  --doc                # Include documentation strings in output\n"
            "  --lang               # Language compliance (ecma|plus|fixed)\n"
            "  --empty              # Create empty interpreter\n"
            "  --endian big|little  # Set the target byte order\n"
            "  --merge              # Merge dependent input modules into the output\n"
            "  --noout              # Do not generate any output\n"
            "  --optimize level     # Set optimization level (0-9)\n"
            "  --out filename       # Name a single output module (default: \"default.mod\")\n"
            "  --parse              # Just parse source. No output\n"
            "  --searchpath ejsPath # Module search path\n"
            "  --standard           # Default compilation mode to standard (default)\n"
            "  --strict             # Default compilation mode to strict\n"
#if FUTURE
            "  --sign certFile      # Sign the module file (not implemented) \n"
            "  --strip              # Strip all symbolic names (Can't import)\n"
            "  --tabwidth           # Tab width for '^' error reporting\n"
#endif
            "  --use 'module, ...'  # List of modules to pre-load\n"
            "  --version            # Emit the compiler version information\n"
            "  --warn level         # Set the warning message level (0-9)\n\n",
            mpr->name);
        return -1;
    }

    /*
     *  Need an interpreter when compiling
     */
    vmService = ejsCreateService(mpr);
    if (vmService == 0) {
        return MPR_ERR_NO_MEMORY;
    }

    if (searchPath) {
        ejsSetSearchPath(vmService, searchPath);
    }

    if (empty) {
        ejsFlags |= EJS_FLAG_EMPTY;
    }
    if (doc) {
        ejsFlags |= EJS_FLAG_DOC;
    }

    ejs = ejsCreate(vmService, NULL, ejsFlags);
    if (ejs == 0) {
        return MPR_ERR_NO_MEMORY;
    }

    ecFlags = 0;
    ecFlags |= (bind) ? EC_FLAGS_BIND_GLOBALS: 0;
    ecFlags |= (debug) ? EC_FLAGS_DEBUG: 0;
    ecFlags |= (empty) ? EC_FLAGS_EMPTY: 0;
    ecFlags |= (merge) ? EC_FLAGS_MERGE: 0;
    ecFlags |= (nobind) ? EC_FLAGS_NO_BIND: 0;
    ecFlags |= (noout) ? EC_FLAGS_NO_OUT: 0;
    ecFlags |= (parseOnly) ? EC_FLAGS_PARSE_ONLY: 0;

    cp = ecCreateCompiler(ejs, ecFlags, lang);
    if (cp == 0) {
        return MPR_ERR_NO_MEMORY;
    }

    ecSetHostEndian(cp, endian);
    ecSetOptimizeLevel(cp, optimizeLevel);
    ecSetWarnLevel(cp, warnLevel);
    ecSetDefaultMode(cp, compilerMode);
    ecSetTabWidth(cp, tabWidth);
    ecSetOutputFile(cp, outputFile);
    ecSetCertFile(cp, certFile);

    if (preloadModules(cp, useModules) < 0) {
        return EJS_ERR;
    }

    if (nextArg < argc) {
        /*
         *  Compile the source files supplied on the command line. This will compile in-memory and
         *  optionally also save to module files.
         */
        if (ecCompile(cp, argc - nextArg, &argv[nextArg], 0) < 0) {
            err++;
        }
    }

    if (cp->errorCount > 0) {
        err++;
    }

    mprFree(mpr);
    return err;
}


/*
 *  This can't be put into ejsCommon because all-in-one builds don't include ec.h which is required by this function.
 */
static int preloadModules(EcCompiler *cp, MprList *modules)
{
    cchar   *name;
    int     next;

    for (next = 0; modules && (name = (cchar*) mprGetNextItem(modules, &next)) != 0; ) {
        if (ejsLoadModule(cp->ejs, name, NULL, NULL, EJS_MODULE_DONT_INIT) < 0) {
            return EJS_ERR;
        }
    }
    cp->useModules = modules;
    return 0;
}


/*
 *  @copy   default
 *
 *  Copyright (c) Embedthis Software LLC, 2003-2009. All Rights Reserved.
 *  Copyright (c) Michael O'Brien, 1993-2009. All Rights Reserved.
 *
 *  This software is distributed under commercial and open source licenses.
 *  You may use the GPL open source license described below or you may acquire
 *  a commercial license from Embedthis Software. You agree to be fully bound
 *  by the terms of either license. Consult the LICENSE.TXT distributed with
 *  this software for full details.
 *
 *  This software is open source; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by the
 *  Free Software Foundation; either version 2 of the License, or (at your
 *  option) any later version. See the GNU General Public License for more
 *  details at: http://www.embedthis.com/downloads/gplLicense.html
 *
 *  This program is distributed WITHOUT ANY WARRANTY; without even the
 *  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 *  This GPL license does NOT permit incorporating this software into
 *  proprietary programs. If you are unable to comply with the GPL, you must
 *  acquire a commercial license to use this software. Commercial licenses
 *  for this software and support services are available from Embedthis
 *  Software at http://www.embedthis.com
 *
 *  @end
 */
